File: frontend/src/services/notificationService.ts
/**
 * Servicio para gestión de notificaciones conectado al backend
 */
import axios from 'axios';
import { buildApiUrl, getAuthHeaders } from '../config/api.config';

interface APIError {
  response?: {
    status: number;
    data: {
      detail: string;
    };
  };
}

export interface Notification {
  id: string;
  recipient_id: string;
  sender_id?: string;
  type: string;
  title: string;
  message: string;
  status: 'unread' | 'read' | 'archived';
  metadata: Record<string, any>;
  action_url?: string;
  action_label?: string;
  created_at: string;
  read_at?: string;
  archived_at?: string;
}

export interface NotificationListResponse {
  notifications: Notification[];
  total: number;
  unread_count: number;
  page: number;
  per_page: number;
}

export interface CreateNotificationData {
  recipient_id: string;
  sender_id?: string;
  type: string;
  title: string;
  message: string;
  metadata?: Record<string, any>;
  action_url?: string;
  action_label?: string;
}

export interface BulkNotificationData {
  recipient_ids: string[];
  sender_id?: string;
  type: string;
  title: string;
  message: string;
  metadata?: Record<string, any>;
  action_url?: string;
  action_label?: string;
}

class NotificationService {
  private readonly baseUrl = '/notifications';

  /**
   * Obtiene las notificaciones del usuario actual
   */
  async getUserNotifications(
    status?: 'unread' | 'read' | 'archived',
    page: number = 1,
    per_page: number = 20
  ): Promise<NotificationListResponse> {
    try {
      console.log('🔍 [notificationService] Fetching user notifications:', { status, page, per_page });
      const params = new URLSearchParams({
        page: page.toString(),
        per_page: per_page.toString()
      });

      if (status) {
        params.append('status', status);
      }

      const response = await axios.get(buildApiUrl(`${this.baseUrl}/?${params}`), { headers: getAuthHeaders() });
      console.log('✅ [notificationService] Notifications loaded:', response.data?.notifications?.length || 0);
      return response.data;
    } catch (error) {
      const apiError = error as APIError;
      console.error('❌ [notificationService] Get notifications error:', apiError);
      const errorMessage = apiError.response?.data?.detail || 'Error fetching notifications';
      throw new Error(errorMessage);
    }
  }

  /**
   * Obtiene el número de notificaciones no leídas
   */
  async getUnreadCount(): Promise<number> {
    try {
      console.log('🔍 [notificationService] Fetching unread count');
      const response = await axios.get(buildApiUrl(`${this.baseUrl}/unread/count`), { headers: getAuthHeaders() });
      const count = response.data?.unread_count || 0;
      console.log('✅ [notificationService] Unread count:', count);
      return count;
    } catch (error) {
      const apiError = error as APIError;
      console.error('❌ [notificationService] Get unread count error:', apiError);
      const errorMessage = apiError.response?.data?.detail || 'Error fetching unread count';
      throw new Error(errorMessage);
    }
  }

  /**
   * Marca una notificación como leída
   */
  async markAsRead(notificationId: string): Promise<void> {
    try {
      console.log('🔍 [notificationService] Marking notification as read:', notificationId);
      await axios.patch(buildApiUrl(`${this.baseUrl}/${notificationId}/read`), {}, { headers: getAuthHeaders() });
      console.log('✅ [notificationService] Notification marked as read:', notificationId);
    } catch (error) {
      const apiError = error as APIError;
      console.error('❌ [notificationService] Mark as read error:', apiError);
      const errorMessage = apiError.response?.data?.detail || 'Error marking notification as read';
      throw new Error(errorMessage);
    }
  }

  /**
   * Marca todas las notificaciones como leídas
   */
  async markAllAsRead(): Promise<number> {
    try {
      console.log('🔍 [notificationService] Marking all notifications as read');
      const response = await axios.patch(buildApiUrl(`${this.baseUrl}/mark-all-read`), {}, { headers: getAuthHeaders() });
      const count = response.data?.message?.match(/\d+/)?.[0] || 0;
      console.log('✅ [notificationService] All notifications marked as read, count:', count);
      return count;
    } catch (error) {
      const apiError = error as APIError;
      console.error('❌ [notificationService] Mark all as read error:', apiError);
      const errorMessage = apiError.response?.data?.detail || 'Error marking all notifications as read';
      throw new Error(errorMessage);
    }
  }

  /**
   * Elimina una notificación
   */
  async deleteNotification(notificationId: string): Promise<void> {
    try {
      console.log('🔍 [notificationService] Deleting notification:', notificationId);
      await axios.delete(buildApiUrl(`${this.baseUrl}/${notificationId}`), { headers: getAuthHeaders() });
      console.log('✅ [notificationService] Notification deleted:', notificationId);
    } catch (error) {
      const apiError = error as APIError;
      console.error('❌ [notificationService] Delete notification error:', apiError);
      const errorMessage = apiError.response?.data?.detail || 'Error deleting notification';
      throw new Error(errorMessage);
    }
  }

  /**
   * Archiva una notificación
   */
  async archiveNotification(notificationId: string): Promise<void> {
    try {
      console.log('🔍 [notificationService] Archiving notification:', notificationId);
      await axios.patch(buildApiUrl(`${this.baseUrl}/${notificationId}/archive`), {}, { headers: getAuthHeaders() });
      console.log('✅ [notificationService] Notification archived:', notificationId);
    } catch (error) {
      const apiError = error as APIError;
      console.error('❌ [notificationService] Archive notification error:', apiError);
      const errorMessage = apiError.response?.data?.detail || 'Error archiving notification';
      throw new Error(errorMessage);
    }
  }

  /**
   * Crea una nueva notificación (solo admins)
   */
  async createNotification(data: CreateNotificationData): Promise<Notification> {
    try {
      console.log('🔍 [notificationService] Creating notification:', data.title);
      const response = await axios.post(buildApiUrl(this.baseUrl), data, { headers: getAuthHeaders() });
      console.log('✅ [notificationService] Notification created:', response.data?.id);
      return response.data;
    } catch (error) {
      const apiError = error as APIError;
      console.error('❌ [notificationService] Create notification error:', apiError);
      const errorMessage = apiError.response?.data?.detail || 'Error creating notification';
      throw new Error(errorMessage);
    }
  }

  /**
   * Crea múltiples notificaciones en lote (solo admins)
   */
  async createBulkNotifications(data: BulkNotificationData): Promise<Notification[]> {
    try {
      console.log('🔍 [notificationService] Creating bulk notifications:', data.recipient_ids.length, 'recipients');
      const response = await axios.post(buildApiUrl(`${this.baseUrl}/bulk`), data, { headers: getAuthHeaders() });
      console.log('✅ [notificationService] Bulk notifications created:', response.data?.length || 0);
      return response.data;
    } catch (error) {
      const apiError = error as APIError;
      console.error('❌ [notificationService] Create bulk notifications error:', apiError);
      const errorMessage = apiError.response?.data?.detail || 'Error creating bulk notifications';
      throw new Error(errorMessage);
    }
  }

  /**
   * Notifica completación de curso al instructor
   */
  async notifyCourseCompletion(
    studentId: string,
    courseId: string,
    courseTitle: string,
    instructorId: string
  ): Promise<void> {
    await axios.post(buildApiUrl(`${this.baseUrl}/course-completion`), {
      student_id: studentId,
      course_id: courseId,
      course_title: courseTitle,
    }, { headers: getAuthHeaders() });
  }

  /**
   * Notifica progreso significativo al instructor
   */
  async notifyCourseProgress(
    studentId: string,
    courseId: string,
    courseTitle: string,
    instructorId: string,
    progressPercentage: number
  ): Promise<void> {
    await axios.post(buildApiUrl(`${this.baseUrl}/course-progress`), {
      student_id: studentId,
      course_id: courseId,
      course_title: courseTitle,
      instructor_id: instructorId,
      progress_percentage: progressPercentage
    }, { headers: getAuthHeaders() });
  }

  /**
   * Notifica nueva inscripción al instructor
   */
  async notifyNewEnrollment(
    studentId: string,
    courseId: string,
    courseTitle: string,
    instructorId: string,
    studentName: string
  ): Promise<void> {
    await axios.post(buildApiUrl(`${this.baseUrl}/enrollment`), {
      student_id: studentId,
      course_id: courseId,
      course_title: courseTitle,
      instructor_id: instructorId,
      student_name: studentName
    }, { headers: getAuthHeaders() });
  }

  /**
   * Notifica completación de quiz al instructor
   */
  async notifyQuizCompletion(
    studentId: string,
    quizId: string,
    quizTitle: string,
    courseId: string,
    instructorId: string,
    score: number
  ): Promise<void> {
    await axios.post(buildApiUrl(`${this.baseUrl}/quiz-completion`), {
      student_id: studentId,
      quiz_id: quizId,
      quiz_title: quizTitle,
      course_id: courseId,
      instructor_id: instructorId,
      score: score
    }, { headers: getAuthHeaders() });
  }

  /**
   * Formatea el timestamp de una notificación
   */
  formatTimestamp(timestamp: string): string {
    const now = new Date();
    const notificationTime = new Date(timestamp);
    const diffMs = now.getTime() - notificationTime.getTime();
    const diffMinutes = Math.floor(diffMs / (1000 * 60));
    const diffHours = Math.floor(diffMs / (1000 * 60 * 60));
    const diffDays = Math.floor(diffMs / (1000 * 60 * 60 * 24));

    if (diffMinutes < 1) return 'Ahora';
    if (diffMinutes < 60) return `${diffMinutes}m`;
    if (diffHours < 24) return `${diffHours}h`;
    return `${diffDays}d`;
  }

  /**
   * Obtiene el icono según el tipo de notificación
   */
  getNotificationIcon(type: string): string {
    const iconMap: Record<string, string> = {
      'course_completion': '🎓',
      'course_progress': '📈',
      'new_course': '📚',
      'enrollment': '👥',
      'quiz_completed': '✅',
      'assignment_submitted': '📝',
      'certificate_awarded': '🏆',
      'course_published': '🚀',
      'lesson_completed': '✏️',
      'system_update': '🔔'
    };
    return iconMap[type] || '🔔';
  }

  /**
   * Convierte tipo de backend a tipo de frontend
   */
  mapNotificationType(backendType: string): 'success' | 'warning' | 'error' | 'info' {
    const typeMap: Record<string, 'success' | 'warning' | 'error' | 'info'> = {
      'course_completion': 'success',
      'course_progress': 'info',
      'new_course': 'info',
      'enrollment': 'success',
      'quiz_completed': 'success',
      'assignment_submitted': 'info',
      'certificate_awarded': 'success',
      'course_published': 'info',
      'lesson_completed': 'success',
      'system_update': 'warning'
    };
    return typeMap[backendType] || 'info';
  }
}

export const notificationService = new NotificationService();

File: backend_python_legacy/app/application/services/notification_service.py
"""
Servicio de aplicación para notificaciones
"""
from typing import List, Optional, Dict, Any
from datetime import datetime

from ...domain.repositories.notification_repository import NotificationRepository
from ...domain.entities.notification import Notification, NotificationType, NotificationStatus
from ..dtos.notification_dto import (
    NotificationCreate, NotificationUpdate, NotificationResponse,
    BulkNotificationCreate, NotificationFilters
)


class NotificationService:
    """Servicio para gestión de notificaciones"""

    def __init__(self, notification_repository: NotificationRepository):
        self.notification_repository = notification_repository

    async def create_notification(self, notification_data: NotificationCreate) -> NotificationResponse:
        """Crea una nueva notificación"""
        notification = Notification(
            recipient_id=notification_data.recipient_id,
            sender_id=notification_data.sender_id,
            type=notification_data.type,
            title=notification_data.title,
            message=notification_data.message,
            metadata=notification_data.metadata or {},
            action_url=notification_data.action_url,
            action_label=notification_data.action_label
        )

        created_notification = await self.notification_repository.create(notification)
        return self._to_response(created_notification)

    async def bulk_create_notifications(self, bulk_data: BulkNotificationCreate) -> List[NotificationResponse]:
        """Crea múltiples notificaciones en lote"""
        notifications = []
        for recipient_id in bulk_data.recipient_ids:
            notification = Notification(
                recipient_id=recipient_id,
                sender_id=bulk_data.sender_id,
                type=bulk_data.type,
                title=bulk_data.title,
                message=bulk_data.message,
                metadata=bulk_data.metadata or {},
                action_url=bulk_data.action_url,
                action_label=bulk_data.action_label
            )
            notifications.append(notification)

        created_notifications = await self.notification_repository.bulk_create(notifications)
        return [self._to_response(notif) for notif in created_notifications]

    async def get_user_notifications(
        self, 
        user_id: str, 
        status: Optional[NotificationStatus] = None,
        limit: int = 50,
        skip: int = 0
    ) -> List[NotificationResponse]:
        """Obtiene notificaciones de un usuario"""
        notifications = await self.notification_repository.get_by_recipient(
            user_id, status, limit, skip
        )
        return [self._to_response(notif) for notif in notifications]

    async def get_unread_count(self, user_id: str) -> int:
        """Obtiene el número de notificaciones no leídas"""
        return await self.notification_repository.get_unread_count(user_id)

    async def mark_as_read(self, notification_id: str) -> bool:
        """Marca una notificación como leída"""
        return await self.notification_repository.mark_as_read(notification_id)

    async def mark_all_as_read(self, user_id: str) -> int:
        """Marca todas las notificaciones como leídas"""
        return await self.notification_repository.mark_all_as_read(user_id)

    async def delete_notification(self, notification_id: str) -> bool:
        """Elimina una notificación"""
        return await self.notification_repository.delete(notification_id)

    async def archive_notification(self, notification_id: str) -> bool:
        """Archiva una notificación"""
        return await self.notification_repository.archive_notification(notification_id)

    # Notificaciones específicas del dominio LMS

    async def notify_course_completion(
        self, 
        student_id: str, 
        course_id: str, 
        course_title: str,
        instructor_id: str
    ) -> None:
        """Notifica completación de curso al instructor"""
        notification = NotificationCreate(
            recipient_id=instructor_id,
            sender_id=student_id,
            type=NotificationType.COURSE_COMPLETION,
            title="Estudiante completó curso",
            message=f"Un estudiante ha completado exitosamente el curso '{course_title}'",
            metadata={
                "course_id": course_id,
                "student_id": student_id,
                "completion_date": datetime.utcnow().isoformat()
            },
            action_url=f"/platform/admin/courses/{course_id}/students",
            action_label="Ver detalles"
        )
        await self.create_notification(notification)

    async def notify_course_progress(
        self, 
        student_id: str, 
        course_id: str, 
        course_title: str,
        instructor_id: str,
        progress_percentage: int
    ) -> None:
        """Notifica progreso significativo del curso al instructor"""
        if progress_percentage >= 80:
            notification = NotificationCreate(
                recipient_id=instructor_id,
                sender_id=student_id,
                type=NotificationType.COURSE_PROGRESS,
                title="Progreso significativo de estudiante",
                message=f"Un estudiante ha alcanzado {progress_percentage}% en el curso '{course_title}'",
                metadata={
                    "course_id": course_id,
                    "student_id": student_id,
                    "progress": progress_percentage
                },
                action_url=f"/platform/admin/courses/{course_id}/students",
                action_label="Ver progreso"
            )
            await self.create_notification(notification)

    async def notify_new_enrollment(
        self, 
        student_id: str, 
        course_id: str, 
        course_title: str,
        instructor_id: str,
        student_name: str
    ) -> None:
        """Notifica nueva inscripción al instructor"""
        notification = NotificationCreate(
            recipient_id=instructor_id,
            sender_id=student_id,
            type=NotificationType.ENROLLMENT,
            title="Nueva inscripción en tu curso",
            message=f"{student_name} se ha inscrito en tu curso '{course_title}'",
            metadata={
                "course_id": course_id,
                "student_id": student_id,
                "enrollment_date": datetime.utcnow().isoformat()
            },
            action_url=f"/platform/admin/courses/{course_id}/students",
            action_label="Ver estudiantes"
        )
        await self.create_notification(notification)

    async def notify_new_course_to_all_students(
        self, 
        course_id: str, 
        course_title: str,
        instructor_name: str,
        student_ids: List[str]
    ) -> None:
        """Notifica nuevo curso a todos los estudiantes"""
        bulk_notification = BulkNotificationCreate(
            recipient_ids=student_ids,
            type=NotificationType.NEW_COURSE,
            title="Nuevo curso disponible",
            message=f"Se ha publicado un nuevo curso: '{course_title}' por {instructor_name}",
            metadata={
                "course_id": course_id,
                "publication_date": datetime.utcnow().isoformat()
            },
            action_url=f"/platform/courses/{course_id}",
            action_label="Ver curso"
        )
        await self.bulk_create_notifications(bulk_notification)

    async def notify_quiz_completion(
        self, 
        student_id: str, 
        quiz_id: str, 
        quiz_title: str,
        course_id: str,
        instructor_id: str,
        score: float
    ) -> None:
        """Notifica completación de quiz al instructor"""
        notification = NotificationCreate(
            recipient_id=instructor_id,
            sender_id=student_id,
            type=NotificationType.QUIZ_COMPLETED,
            title="Quiz completado por estudiante",
            message=f"Un estudiante completó el quiz '{quiz_title}' con puntaje: {score}%",
            metadata={
                "quiz_id": quiz_id,
                "course_id": course_id,
                "student_id": student_id,
                "score": score
            },
            action_url=f"/platform/admin/quizzes/{quiz_id}/statistics"
        )
        
        await self.create_notification(notification)
    
    def _to_response(self, notification: Notification) -> NotificationResponse:
        """Convierte entidad a DTO de respuesta"""
        return NotificationResponse(
            id=notification.id,
            recipient_id=notification.recipient_id,
            sender_id=notification.sender_id,
            type=notification.type.value,
            title=notification.title,
            message=notification.message,
            status=notification.status.value,
            metadata=notification.metadata,
            action_url=notification.action_url,
            action_label=notification.action_label,
            created_at=notification.created_at,
            read_at=notification.read_at,
            archived_at=notification.archived_at
        )
